Improve BYOD for HPO and Ablation pipelines

mali-git opened the pull request:
BYOD means *bring your own data*.

Closes #30

This PR does the following:

1. Reorganizes ablation code to be more reusable
2. Adds a tutorial for ablation studies
3. Enables definition of triples factories by `str` in the pipeline, HPO pipeline, and ablation pipeline
4. Updates tutorials for pipeline/HPO for usage of custom triples

Before, `pykeen experiments ablation` could only be run with PyKEEN internal datasets. Now, this PR enables running the function with user provided datasets. For example, in a configuration file the user can provide the training, validation, and test data as follows:

```python
"datasets": [
      {
        "training": "/path/to/train.txt",
        "testing": "/path/to/test.txt",
        "validation": "/path/to/valid.txt"
      }
]
```

Before, the elements in `datasets` needed to be dataset-names (strings) of integrated datasets. Now, the user can also provide the paths to their own data by means of a dictionary (both cases can be mixed).

mali-git commented:
Thanks @annabreit!

cthoyt commented:
Likely my fault. Thanks for catching this

mali-git committed:
Allow pipeline's triples_factory-argument to be a path.

mali-git committed:
Handle case in which path to triples are provided.

mali-git committed:
Extract triples_factory key-value pairs from config which can represent paths

mali-git committed:
Enable processing of PyKEEN internal dataset and user provided dataset

mali-git committed:
Pass flake8

mali-git committed:
Fix error message

mali-git committed:
Fix error message

mali-git committed:
Fix error message

cthoyt committed:
Improve pipeline documentation

[skip ci]

cthoyt commented on src/pykeen/datasets/__init__.py:
```diff
@@ -84,6 +84,15 @@ def get_dataset(
             **(dataset_kwargs or {})
         )
         return dataset_instance.factories
+    elif training_triples_factory is not None and testing_triples_factory is not None:
```
this check needs to be reorganized

cthoyt committed:
Reorganize dataset passing

- Add EagerDataset for wrapping 3 triples factories
- Return DataSet instance from get_dataset (streamlined type checking and instiation)
- Update logging

cthoyt commented on src/pykeen/ablation/ablation.py:
```diff
@@ -97,7 +98,16 @@ def _set_arguments(key: str, value: str) -> None:
+            hpo_config['dataset'] = dataset
+        elif isinstance(dataset, dict):
+            # Training, test, and validation paths are provided
+            hpo_config['training_triples_factory'] = dataset['training_triples_factory']
```
should we remove the suffix "triples_factory" since this can also be a string now?

mali-git replied:
Good idea!
I made the changes in 4a892c91fe8d3723fceb33e9fed7d42e2171be00. Please note that I updated the `pipeline` accordingly.

cthoyt committed:
Update docs and fixmes

[skip ci]

cthoyt committed:
Update tutorial

make it more clear that this is also applicable for HPO

cthoyt committed:
Update pipeline.py

[skip ci]

cthoyt commented on src/pykeen/hpo/hpo.py:
```diff
@@ -579,6 +582,7 @@ def hpo_pipeline(
     study.set_user_attr('pykeen_version', get_version())
     study.set_user_attr('pykeen_git_hash', get_git_hash())
     # 1. Dataset
+    # FIXME must be addressed before merging https://github.com/pykeen/pykeen/pull/54
```
FIXME!!!

cthoyt replied:
@cthoyt todo

cthoyt replied:
@cthoyt yup

cthoyt committed:
Merge branch 'master' into hpo_helper_fcts_customized_data

cthoyt committed:
Add documentation todos for master mehdi

@mali-git check

cthoyt committed:
Reorganize code + docs for ablation studies

cthoyt commented on src/pykeen/ablation/__init__.py:
```diff
@@ -1,10 +1,31 @@
+  killed by your dead wife, then :func:`pykeen.ablation.ablation_pipeline` is washing up on
+  a beach." -Christopher Nolan
+
+TODO for Mehdi!
```
@mali-git TODO

mali-git replied:
Done in 7e1bafa0ca149b941c001c80fc24f0048a4c8682 - 8ad9dfe31c7f590b6b93e865c18e7061f6935816.

cthoyt commented on src/pykeen/ablation/ablation.py:
```diff
@@ -6,29 +6,81 @@
+    move_to_cpu: bool,
+    discard_replicates: bool,
+) -> None:
+    """Generate a set of HPO configurations.
```
@mali-git also need better docs here (and other functions in this file)

mali-git replied:
Done in 78971d6bd839176e9c1dc758a9b5c9b0e2fd08e3.

mali-git committed:
Remove triples_factory suffix

mali-git committed:
Fix loading of config

mali-git committed:
Rename argument and adapt type annotation

mali-git committed:
Add HPO tutorial

mali-git committed:
Adapt tutorial

mali-git committed:
Adapt tutorial

mali-git committed:
Provide tolerance value to unit test

mali-git committed:
Fix path in tutorial

mali-git committed:
Extend documentation of ablation functions.
Add missing type annotations

mali-git committed:
Pass flake8

mali-git committed:
Adapt indention

cthoyt commented on src/pykeen/ablation/__init__.py:
```diff
@@ -1,10 +1,270 @@
+Define Ablation Dictionary
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+>>> ablation = {}
+
```
the way you wrote this is really counterintuitive. Each code block should be completely executable. use the 

```
.. code-block:: python

   ...
```

to make it easier to put more code together and use python commenting

cthoyt commented on src/pykeen/ablation/__init__.py:
```diff
@@ -1,10 +1,270 @@
-from .ablation import prepare_ablation, prepare_ablation_from_config
+  "If :func:`pykeen.pipeline.pipeline` were making a maze, :func:`pykeen.hpo.hpo_pipeline` were getting
+  killed by your dead wife, then :func:`pykeen.ablation.ablation_pipeline` is washing up on
+  a beach." -Christopher Nolan
```
made up quote should be removed... and I should be put in time out

mali-git committed:
Adapt tutorial

mali-git committed:
Remove placeholder and fix grammar

mali-git committed:
Remove comment

cthoyt committed:
Merge branch 'master' into hpo_helper_fcts_customized_data

cthoyt committed:
Fix spacing

cthoyt committed:
Add proper indentation

cthoyt committed:
Merge branch 'master' into hpo_helper_fcts_customized_data

cthoyt committed:
Add promprt

[skip ci]

cthoyt commented on src/pykeen/ablation/ablation.py:
```diff
@@ -6,29 +6,102 @@
+
+    directories = prepare_ablation_from_config(config=config, directory=directory, save_artifacts=save_artifacts)
+    if dry_run:
+        return sys.exit(0)
```
no need to exit. just return. what if there's other code that should run after?

cthoyt committed:
Merge branch 'master' into hpo_helper_fcts_customized_data

cthoyt committed:
Add simple example

cthoyt committed:
Merge branch 'master' into hpo_helper_fcts_customized_data

cthoyt committed:
Pass flake8

mali-git committed:
Adapt type annotation

mali-git committed:
Fix example

lvermue committed:
Merge branch 'master' into hpo_helper_fcts_customized_data

cthoyt committed:
Merge branch 'master' into hpo_helper_fcts_customized_data

cthoyt committed:
Update ablation.py

cthoyt committed:
Remove unfinished ablation tutorial

cthoyt committed:
Update __init__.py

cthoyt committed:
Minimal solution for dataset name problem

mberr commented on src/pykeen/ablation/ablation.py:
```diff
@@ -6,29 +6,172 @@
+    :param move_to_cpu: Defines, whether a replicate of the best model should be moved to CPU.
+     We recommend to set this flag to 'True' to avoid unnecessary GPU usage.
+    :param discard_replicates: Defines, whether the best model should be discarded after training and evaluation.
+    """
```
there are some undocumented parameters, e.g. `model_to_trainer_to_training_kwargs_ranges`

cthoyt replied:
All ablation-related documentation updates are going to get bumped to https://github.com/pykeen/pykeen/issues/116 for @mali-git to finish, but I will double check missing params in docstrings now

cthoyt replied:
Done in 312a7e4d

cthoyt committed:
Add more docs

cthoyt committed:
Merge branch 'master' into hpo_helper_fcts_customized_data
